docs/standard/native-interop/tutorial-custom-marshaller.md

```diff
@@ -0,0 +1,278 @@
+
+```csharp
+// extern "C" DLL_EXPORT void STDMETHODCALLTYPE PrintString(char32_t* chars);
+[LibraryImport(LibName)]
+internal static partial void PrintString([MarshalUsing(typeof(Utf32StringMarshaller))] string s);
+```
+
+## Customize marshalling for a user-defined type
+
+Marshalling a user-defined type requires defining not only the marshalling logic, but also the type in C# to marshal to/from. Let's recall the native type we are trying to marshal and define what it would ideally look like in C#.
+
+```cpp
+struct error_data
+{
+    int code;
+    bool is_fatal_error;
+    char32_t* message;    /* UTF-32 encoded string */
+};
+```
+
+An `int` is the same size in both modern C++ and in .NET. A `bool` is the canonical example for a boolean value in .NET. Building on top of `Utf32StringMarshaller`, we can marshal `char32_t*` as a .NET `string`. Accounting for .NET style the result is the following definition in C#.
+
+```csharp
+struct ErrorData
+{
+    public int Code;
+    public bool IsFatalError;
+    public string? Message;
+}
+```
+
+Following the naming pattern, we will name the marshaller `ErrorDataMarshaller`. Instead of creating a default marshaller, we will only define marshallers for some modes. This will demonstrate the case where if a mode isn't provided, the source generator will fail. Let's start with the "in" direction.
+
+```csharp
+namespace CustomMarshalling
+{
+    [CustomMarshaller(typeof(ErrorData), MarshalMode.ManagedToUnmanagedIn, typeof(ErrorDataMarshaller))]
+    internal static unsafe class ErrorDataMarshaller
+    {
+        // Unmanaged representation of ErrorData.
+        // Should mimic the unmanaged error_data type at a binary level.
+        internal struct ErrorDataUnmanaged
+        {
+            public int Code;        // .NET doesn't support less than 32-bit, so int is 32-bit.
+            public byte IsFatal;    // The C++ bool is defined as a single byte.
+            public uint* Message;   // This could be as simple as a void*, but uint* is closer.
+        }
+
+        public static ErrorDataUnmanaged ConvertToUnmanaged(ErrorData managed)
+            => throw new NotImplementedException();
+
+        public static void Free(ErrorDataUnmanaged unmanaged)
+            => throw new NotImplementedException();
+    }
+}
+```
+
+The conversion from an `ErrorData` to an `ErrorDataUnmanaged`, that will mimic the shape of the unmanaged type is now trivial with `Utf32StringMarshaller`. Note this is a "stateless" marshaller because the marshaller itself only consists of `static` functions.
```

Comment:
```suggestion
The conversion from an `ErrorData` to an `ErrorDataUnmanaged`, which will mimic the shape of the unmanaged type, is now trivial with `Utf32StringMarshaller`. This is a "stateless" marshaller because the marshaller itself only consists of `static` functions.
```